From 9bc9d3a4daf0bdebb1c3e9c7efeb54eb8e1c4e22 Mon Sep 17 00:00:00 2001
From: Tobias Schramm <t.schramm@t-sys.eu>
Date: Mon, 15 Mar 2021 12:23:41 +0100
Subject: [PATCH 4/6] arm: sun8i: add support for OtterCast Amp

The OtterCast Amp is a S3 based SBC for audio streaming. It integrates two
TAS5825 class D amplifiers with builtin codec, fast ethernet via the internal
MAC and PHY and dual-band WLAN/BT through an AP6255 SDIO module.
---
 arch/arm/dts/Makefile                   |   1 +
 arch/arm/dts/sun8i-s3-ottercast-amp.dts | 154 ++++++++++++++++++++++++
 configs/ottercast_amp_defconfig         |  13 ++
 3 files changed, 168 insertions(+)
 create mode 100644 arch/arm/dts/sun8i-s3-ottercast-amp.dts
 create mode 100644 configs/ottercast_amp_defconfig

diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index 9aa29d2463..f3b0b53f06 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -598,6 +598,7 @@ dtb-$(CONFIG_MACH_SUN8I_R40) += \
 	sun8i-r40-bananapi-m2-ultra.dtb \
 	sun8i-v40-bananapi-m2-berry.dtb
 dtb-$(CONFIG_MACH_SUN8I_V3S) += \
+	sun8i-s3-ottercast-amp.dtb \
 	sun8i-s3-ottercast-audio-v2.dtb \
 	sun8i-s3-pinecube.dtb \
 	sun8i-v3s-licheepi-zero.dtb
diff --git a/arch/arm/dts/sun8i-s3-ottercast-amp.dts b/arch/arm/dts/sun8i-s3-ottercast-amp.dts
new file mode 100644
index 0000000000..ed86554f30
--- /dev/null
+++ b/arch/arm/dts/sun8i-s3-ottercast-amp.dts
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR X11)
+/*
+ * Copyright 2021 Tobias Schramm <t.schramm@t-sys.eu>
+ */
+
+/dts-v1/;
+#include "sun8i-v3.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	model = "OtterCast Amp";
+	compatible = "otter,ottercast-amp", "allwinner,sun8i-s3";
+
+	aliases {
+		serial0 = &uart1;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led1 {
+			label = "led:usr";
+			gpios = <&pio 4 0 GPIO_ACTIVE_HIGH>; /* PE0 */
+		};
+
+		led2 {
+			label = "led:pwr";
+			gpios = <&pio 4 1 GPIO_ACTIVE_HIGH>; /* PE1 */
+			default-state = "on";
+		};
+	};
+
+	/*
+	 * Power tree
+	 * This is a virtual regulator, root of the power tree
+	 */
+
+	vcc_vbus: vcc-vbus {
+		compatible = "regulator-fixed";
+		regulator-always-on;
+		regulator-name = "vcc-vbus";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+	};
+
+	/* Regulators supplied by vcc_vbus */
+
+	vcc_3v3: vcc-3v3 {
+		compatible = "regulator-fixed";
+		regulator-always-on;
+		regulator-name = "vcc-3v3";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		vin-supply = <&vcc_vbus>;
+	};
+
+	/* Regulators supplied by vcc_3v3 */
+
+	vcc_ddr: vcc-ddr {
+		compatible = "regulator-fixed";
+		regulator-always-on;
+		regulator-name = "vcc-ddr";
+		regulator-min-microvolt = <1500000>;
+		regulator-max-microvolt = <1500000>;
+		vin-supply = <&vcc_3v3>;
+	};
+
+	vcc_1v1: vcc-1v1 {
+		compatible = "regulator-fixed";
+		regulator-always-on;
+		regulator-name = "vcc-1v1";
+		regulator-min-microvolt = <1100000>;
+		regulator-max-microvolt = <1100000>;
+		vin-supply = <&vcc_3v3>;
+	};
+
+	vcc_3v0a: vcc-3v0a {
+		compatible = "regulator-fixed";
+		regulator-always-on;
+		regulator-name = "vcc-3v0a";
+		regulator-min-microvolt = <3000000>;
+		regulator-max-microvolt = <3000000>;
+		vin-supply = <&vcc_3v3>;
+	};
+
+	wifi_pwrseq: wifi_pwrseq {
+		compatible = "mmc-pwrseq-simple";
+		reset-gpios = <&pio 1 11 GPIO_ACTIVE_LOW>; /* PB11 WL-PMU-EN */
+		post-power-on-delay-ms = <200>;
+	};
+};
+
+&emac {
+	phy-handle = <&int_mii_phy>;
+	phy-mode = "mii";
+	status = "okay";
+};
+
+&i2c1 {
+	status = "okay";
+};
+
+&mmc0 {
+	vmmc-supply = <&vcc_3v3>;
+	bus-width = <4>;
+	broken-cd;
+	status = "okay";
+};
+
+&mmc1 {
+	vmmc-supply = <&vcc_3v3>;
+	vqmmc-supply = <&vcc_3v3>;
+	mmc-pwrseq = <&wifi_pwrseq>;
+	bus-width = <4>;
+	non-removable;
+	status = "okay";
+};
+
+&mmc2 {
+	vmmc-supply = <&vcc_3v3>;
+	vqmmc-supply = <&vcc_3v3>;
+	mmc-pwrseq = <&wifi_pwrseq>;
+	bus-width = <8>;
+	non-removable;
+	status = "okay";
+};
+
+&pio {
+	vcc-pd-supply = <&vcc_3v3>;
+	vcc-pe-supply = <&vcc_3v3>;
+};
+
+&uart1 {
+	status = "okay";
+
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart1_pg_pins>;
+};
+
+&uart2 {
+	status = "okay";
+};
+
+&pio {
+	uart1_pg_pins: uart1-pg-pins {
+		pins = "PG6", "PG7";
+		function = "uart1";
+	};
+};
diff --git a/configs/ottercast_amp_defconfig b/configs/ottercast_amp_defconfig
new file mode 100644
index 0000000000..707b83610d
--- /dev/null
+++ b/configs/ottercast_amp_defconfig
@@ -0,0 +1,13 @@
+CONFIG_ARM=y
+CONFIG_ARCH_SUNXI=y
+CONFIG_SPL=y
+CONFIG_MACH_SUN8I_V3S=y
+CONFIG_SUNXI_DRAM_DDR3_1333=y
+CONFIG_DRAM_CLK=504
+CONFIG_DRAM_ODT_EN=y
+CONFIG_DEFAULT_DEVICE_TREE="sun8i-s3-ottercast-amp"
+CONFIG_LED=y
+CONFIG_LED_BLINK=y
+CONFIG_LED_GPIO=y
+# CONFIG_NETDEVICES is not set
+CONFIG_CONS_INDEX=2
-- 
2.30.1

